File: examples/lh-server/typescript/src/10-interrupts/main.ts
import { createTaskWorker, LHConfig, Workflow } from "littlehorse-client";
import { z } from "zod";
import { closeOnShutdown } from "../config.js";

function startCollection(): void {
  console.log("Starting collection of underpants!");
}

function collectUnderpant(underpantOwner: string): string {
  const result = `Successfully collected underpant from ${underpantOwner}`;
  console.log(result);
  return result;
}

function profit(underpants: unknown[]): string {
  const result = `Collected ${underpants.length} underpants!`;
  console.log(result);
  return result;
}

const config = LHConfig.from({});
const client = config.getClient();
const workers = [
  createTaskWorker(startCollection, "start-underpants-collection", config, {
    inputVars: {},
  }),
  createTaskWorker(collectUnderpant, "collect-underpant", config, {
    inputVars: { underpantOwner: z.string() },
    outputSchema: z.string(),
  }),
  createTaskWorker(profit, "profit", config, {
    inputVars: { underpants: z.array(z.unknown()) },
    outputSchema: z.string(),
  }),
];

await client.putExternalEventDef({ name: "underpant-collected", contentType: {} });
await client.putExternalEventDef({
  name: "done-collecting-underpants",
  contentType: {},
});

const workflow = Workflow.newWorkflow("collect-underpants", (wf) => {
  const allUnderpants = wf.declareJsonArr("all-underpants");
  wf.execute("start-underpants-collection");
  wf.waitForEvent("done-collecting-underpants");
  wf.execute("profit", allUnderpants);

  wf.registerInterruptHandler("underpant-collected", (handler) => {
    const interruptContent = handler.declareStr("INPUT");
    allUnderpants.assign(allUnderpants.add(interruptContent));
    handler.execute("collect-underpant", interruptContent);
  });
});

await closeOnShutdown(workers);
await Promise.all(workers.map((worker) => worker.registerTaskDef()));
await workflow.registerWfSpec(config);
console.log("Run with: lhctl run collect-underpants --wfRunId my-wf all-underpants '[]'");
console.log("Then post interrupts with: lhctl postEvent my-wf underpant-collected STR anakin");
console.log("Finish with: lhctl postEvent my-wf done-collecting-underpants");
await Promise.all(workers.map((worker) => worker.start()));

File: examples/lh-server/python/10-interrupts/main.py
import asyncio
import sys

import littlehorse
from littlehorse.config import LHConfig
from littlehorse.model import PutExternalEventDefRequest
from littlehorse.worker import LHTaskWorker
from tasks import profit, ship_item, start_collection
from workflow import UnderpantsWorkflow

TASKS = (
    (start_collection, "start-underpants-collection"),
    (ship_item, "collect-underpant"),
    (profit, "profit"),
)


def register(config: LHConfig) -> None:
    client = config.stub()
    client.PutExternalEventDef(PutExternalEventDefRequest(name="underpant-collected"))
    client.PutExternalEventDef(
        PutExternalEventDefRequest(name="done-collecting-underpants")
    )
    for function, name in TASKS:
        littlehorse.create_task_def(function, name, config)
    littlehorse.create_workflow_spec(UnderpantsWorkflow().get_workflow(), config)


async def workers(config: LHConfig) -> None:
    await littlehorse.start(
        *(LHTaskWorker(function, name, config) for function, name in TASKS)
    )


if __name__ == "__main__":
    config = LHConfig()
    if len(sys.argv) != 2 or sys.argv[1] not in {"register", "workers"}:
        raise SystemExit("Usage: main.py [register|workers]")
    register(config) if sys.argv[1] == "register" else asyncio.run(workers(config))
